hilti/toolchain/src/compiler/cxx/elements.cc

```diff
@@ -21,134 +21,165 @@ static const unsigned int AddSeparatorAfter = (1U << 2U);  // Force adding a sep
 static const unsigned int AddSeparatorBefore = (1U << 4U); // Force adding a separator before block.
 } // namespace flags
 
+static const std::set<std::string_view> reserved_keywords = {
+    "NULL",
+    "_Alignas",
+    "_Alignof",
+    "_Atomic",
+    "_Bool",
+    "_Complex",
+    "_Decimal128",
+    "_Decimal32",
+    "_Decimal64",
+    "_Generic",
+    "_Imaginary",
+    "_Noreturn",
+    "_Pragma",
+    "_Static_assert",
+    "_Thread_local",
+    "alignas",
+    "alignof",
+    "and",
+    "and_eq",
+    "asm",
+    "atomic_cancel",
+    "atomic_commit",
+    "atomic_noexcept",
+    "auto",
+    "bitand",
+    "bitor",
+    "bool",
+    "break",
+    "case",
+    "catch",
+    "char",
+    "char16_t",
+    "char32_t",
+    "char8_t",
+    "class",
+    "co_await",
+    "co_return",
+    "co_yield",
+    "compl",
+    "concept",
+    "const",
+    "const_cast",
+    "consteval",
+    "constexpr",
+    "constinit",
+    "continue",
+    "decltype",
+    "default",
+    "delete",
+    "do",
+    "double",
+    "dynamic_cast",
+    "else",
+    "enum",
+    "explicit",
+    "export",
+    "extern",
+    "false",
+    "float",
+    "for",
+    "fortran",
+    "friend",
+    "goto",
+    "if",
+    "inline",
+    "int",
+    "long",
+    "mutable",
+    "namespace",
+    "new",
+    "noexcept",
+    "not",
+    "not_eq",
+    "nullptr",
+    "operator",
+    "or",
+    "or_eq",
+    "private",
+    "protected",
+    "public",
+    "reflexpr",
+    "register",
+    "reinterpret_cast",
+    "requires",
+    "restrict",
+    "return",
+    "short",
+    "signed",
+    "sizeof",
+    "static",
+    "static_assert",
+    "static_cast",
+    "struct",
+    "switch",
+    "synchronized",
+    "template",
+    "this",
+    "thread_local",
+    "true",
+    "typedef",
+    "typeid",
+    "typename",
+    "union",
+    "unsigned",
+    "using",
+    "virtual",
+    "void",
+    "volatile",
+    "wchar_t",
+    "while",
+    "xor",
+    "xor_eq",
+};
+
+std::optional<std::string> cxx::normalizeID(std::string_view id) {
+    if ( id.empty() )
+        return std::nullopt;
+
+    if ( reserved_keywords.count(id) )
+        return std::string(id) + "_";
+
+    if ( std::all_of(id.begin(), id.end(), [](auto c) { return std::isalnum(c) || c == '_'; }) )
+        // Fast-path: no special-characters, no leading digits.
+        return std::nullopt;
+
+    char buffer[id.size() * 6 + 1]; // max possible size of modified string
+    char* p = buffer;
```

Comment:
It would be great to stay clear of VLAs since they are a non-standard extension, and require us to copy the data to compute the return value.

To do that you could have here

```cpp
    std::string buffer(id.size() * 6 + 1, '\0'); // max possible size of modified string
    char* p = buffer.data();
```

and then at the end

```cpp
    assert(p < buffer.data() + buffer.size());
    auto len = p - buffer.data();
    buffer.erase(len);
    buffer.shrink_to_fit();
    return buffer;
```

Reply:
That at-the-end part doesn't look to appealing to me:  very technical and I'm not sure the shrink-to-fit is better than the copy for the return value. But we can get rid of the VLA by using alloca(), I'll. make that change.